// charto-render/src/GLRenderer.ts
// This file is part of charto-render, copyright (c) 2017 BusFaster Ltd.
// Released under the MIT license, see LICENSE.

import { Point } from 'cgeo';
import { BBox, Layer, LayerFeatures } from 'charto-model';
import { RendererOptions, RendererState }from './RendererState';

export interface GLRendererOptions extends RendererOptions {

	canvas: HTMLCanvasElement;

}

export class GLRendererState extends RendererState {
}

export class GLRenderer {

	protected getContext(options: GLRendererOptions) {
try {
		if(!this.gl) {
			const canvas = document.createElement('canvas');
			canvas.width = options.pixelWidth;
			canvas.height = options.pixelHeight;

			const gl = canvas.getContext('experimental-webgl')!;

			gl.clearColor(0, 0, 0, 0);
			gl.clearDepth(1);

			// gl.enable(gl.DEPTH_TEST);
			// gl.depthFunc(gl.LEQUAL);

			this.canvas = canvas;
			this.gl = gl;

canvas.style.position = 'absolute';
document.body.appendChild(canvas);
console.log(canvas);
		}
} catch(e) { console.error(e); }

		return(this.gl);
	}

	render(options: GLRendererOptions) {
		const state = new GLRendererState(options);
		const gl = this.getContext(options);

		gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
	}

	canvas: HTMLCanvasElement;
	gl: WebGLRenderingContext;

}
