// cgeo-gml/src/ConfigGML.ts
// This file is part of cgeo-gml, copyright (c) 2017 BusFaster Ltd.
// Released under the MIT license, see LICENSE.

import * as cgeo from 'cgeo';
import { OpenToken, TokenElement, ParserConfig } from 'cxml';
import { jsx as gml } from '../dist/gml';

export interface OptionsGML {
	config: ParserConfig;
	flipXY?: boolean;
}

export class ConfigGML implements OptionsGML {

	constructor(options: OptionsGML) {
		this.config = options.config;
		this.flipXY = options.flipXY || false;

		this.config.jsxRegister({ gml }, ({ gml }) => {
			const element = this.elementTbl;
			const member = this.memberTbl;
			this.gml = gml;

			element[cgeo.GeometryKind.multiPoint] = gml.MultiPoint;
			element[cgeo.GeometryKind.multiLineString] = gml.MultiLineString;
			element[cgeo.GeometryKind.multiPolygon] = gml.MultiPolygon;
			element[cgeo.GeometryKind.geometryCollection] = gml.MultiGeometry;
			element[cgeo.GeometryKind.compoundCurve] = gml.CompositeCurve;
			element[cgeo.GeometryKind.multiCurve] = gml.MultiCurve;
			element[cgeo.GeometryKind.multiSurface] = gml.MultiSurface;

			member[cgeo.GeometryKind.multiPoint] = gml.pointMembers;
			member[cgeo.GeometryKind.multiLineString] = gml.lineStringMembers;
			member[cgeo.GeometryKind.multiPolygon] = gml.polygonMembers;
			member[cgeo.GeometryKind.geometryCollection] = gml.geometryMembers;
			member[cgeo.GeometryKind.compoundCurve] = gml.curveMembers;
			member[cgeo.GeometryKind.multiCurve] = gml.curveMembers;
			member[cgeo.GeometryKind.multiSurface] = gml.surfaceMembers;
		});
	}

	config: ParserConfig;
	flipXY: boolean;
	gml: typeof gml;
	elementTbl: OpenToken[] = [];
	memberTbl: OpenToken[] = [];

}
