// cgeo/src/Geometry.ts
// This file is part of cgeo, copyright (c) 2017 BusFaster Ltd.
// Released under the MIT license, see LICENSE.

export function registerType(Type: { new(): Geometry }, kind: GeometryKind, dimension?: number) {
	Type.prototype.kind = kind;
	Type.prototype.dimension = dimension;
	Geometry.typeList[kind] = Type;
}

export enum GeometryKind {
	geometry = 0,
	point = 1,
	lineString = 2,
	polygon = 3,
	multiPoint = 4,
	multiLineString = 5,
	multiPolygon = 6,
	geometryCollection = 7,
	circularString = 8,
	compoundCurve = 9,
	curvePolygon = 10,
	multiCurve = 11,
	multiSurface = 12,
	curve = 13,
	surface = 14,
	polyhedralSurface = 15
}

export abstract class Geometry {

	hasZ() { return(false); }
	hasM() { return(false); }

	kind: GeometryKind;

	/** Dimension of space enclosed by the geometry.
	  * 0 (points), 1 (curves) or 2 (surfaces).
	  * Collections use the largest dimension contained. */
	dimension?: number;
	static typeList: ({ new(): Geometry } | null)[] = [];
	parent?: Geometry;
	id?: number | string;

}

Geometry.prototype.kind = GeometryKind.geometry;
