// cgeo/src/Polygon.ts
// This file is part of cgeo, copyright (c) 2017 BusFaster Ltd.
// Released under the MIT license, see LICENSE.

import { Geometry, GeometryKind, registerType } from './Geometry';
import { GeometryCollection } from './GeometryCollection';
import { LineString } from './LineString';
import { CurveSpec } from './Curve';
import { Surface } from './Surface';
import { initCurves } from './MultiCurve';

export type PolygonSpec = CurveSpec | (CurveSpec | LineString | null | undefined)[];

// GeometryCollection compatibility allows calling methods from MultiCurve.
export class Polygon extends Surface implements GeometryCollection<LineString | null | undefined> {

	constructor(ringList: PolygonSpec = []) {
		super();

		if(this.initProto) this.initProto();

		initCurves(this, ringList, true, false);
	}

	initProto() {
		Polygon.prototype.clone = GeometryCollection.prototype.clone;
		Polygon.prototype.transform = GeometryCollection.prototype.transform;
		Polygon.prototype.hasZ = GeometryCollection.prototype.hasZ;
		Polygon.prototype.hasM = GeometryCollection.prototype.hasM;
		(Polygon as any).initProto = null;
	}

	addChild(child: LineString) { this.childList.push(child); }

	childList: ( LineString | null | undefined )[] = [];

}

registerType(Polygon, GeometryKind.polygon, 2);
