// charto-3d/src/Shader.ts
// This file is part of charto-3d, copyright (C) 2016-2017 BusFaster Ltd.
// Released under the MIT license, see LICENSE.

export const enum Attribute {
	position = 0,
	uv = 1
}

export interface ShaderSpec {
	vertex: string;
	fragment: string;
	attributes: { [ kind: number ]: string };
}

export class Shader {

	constructor(private gl: WebGLRenderingContext, spec: ShaderSpec) {
		const program = gl.createProgram()!;

		gl.attachShader(program, this.compile(spec.vertex, gl.VERTEX_SHADER));
		gl.attachShader(program, this.compile(spec.fragment, gl.FRAGMENT_SHADER));

		for(let kind of Object.keys(spec.attributes)) {
			const num: Attribute = +kind;

			gl.bindAttribLocation(program, num, spec.attributes[num]);
			this.attributeActiveList[num] = true;
		}

		gl.linkProgram(program);

		if(!gl.getProgramParameter(program, gl.LINK_STATUS)) {
			throw(new Error('Shader link error: ' + gl.getProgramInfoLog(program)));
		}

		gl.useProgram(program);
		this.program = program;
	}

	activate(attributeActiveList: boolean[]) {
		const gl = this.gl;
		gl.useProgram(this.program);

		const attributeCount = Math.max(attributeActiveList.length, this.attributeActiveList.length);

		for(let num: Attribute = 0; num < attributeCount; ++num) {
			if(attributeActiveList[num] != this.attributeActiveList[num]) {
				if(attributeActiveList[num]) {
					gl.enableVertexAttribArray(num);
				} else {
					gl.disableVertexAttribArray(num);
				}
			}
		}
	}

	/** @param source Plain-text WebGL shader source code.
	  * @param kind gl.VERTEX_SHADER or gl.FRAGMENT_SHADER. */

	private compile(source: string, kind: number) {
		const gl = this.gl;
		const shader = gl.createShader(kind)!;

		gl.shaderSource(shader, source);
		gl.compileShader(shader);

		if(!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
			throw(new Error('Shader compile error: ' + gl.getShaderInfoLog(shader)));
		}

		return(shader);
	}

	program: WebGLProgram;
	attributeActiveList: boolean[] = [];

}
